src/ir/auth_logic/lowering_ast_datalog.h

```diff
@@ -0,0 +1,212 @@
+        Sign::kPositive);
+
+    Predicate speaker_says_y_can_act_as_x =
+        PushPrin(std::string("says_"), speaker, y_can_act_as_x);
+
+    // This is `speaker says X PredX`
+    Predicate speaker_says_x_pred =
+        PushPrin(std::string("says_"), speaker, main_predicate);
+
+    // This is the full generated rule:
+    // `speaker says Y PredX :-
+    //    speaker says Y canActAs X, speaker says X PredX`
+    return DLIRAssertion(DLIRCondAssertion(
+        generated_lhs, {std::move(speaker_says_y_can_act_as_x),
+                        std::move(speaker_says_x_pred)}));
+  }
+
+  // The second return value represents 0 or 1 newly generated rules, so an
+  // option might seem more intuitive. However, the interface that consumes
+  // this needs to construct a vector anyway.
+  std::pair<Predicate, DLIRAssertion> BaseFactToDLIR(
+      Principal speaker, BaseFact base_fact) {
+    return std::visit(
+        raksha::utils::overloaded{
+            [](Predicate predicate) {
+              Predicate dummy_pred =
+                  Predicate("pred_name", {"arg1"}, kPositive);
+              DLIRAssertion dummy_assertion =
+                  DLIRAssertion(DLIRCondAssertion(dummy_pred, {dummy_pred}));
+              return std::make_pair(predicate, dummy_assertion);
+            },
+            [this, &speaker](Attribute attribute) {
+              return std::make_pair(AttributeToDLIR(attribute),
+                                    SpokenAttributeToDLIR(speaker, attribute));
+            },
+            [](CanActAs canActAs) {
+              Predicate dummy_pred =
+                  Predicate("pred_name", {"arg1"}, kPositive);
+              DLIRAssertion dummy_assertion =
+                  DLIRAssertion(DLIRCondAssertion(dummy_pred, {dummy_pred}));
+              return std::make_pair(dummy_pred, dummy_assertion);
+            }},
+        base_fact.GetValue());
+  }
+
+  // This can result in 0 or more new rules because the translation of
+  // nested canSayFacts might result in more than 1 rule.
+  std::pair<Predicate, std::vector<DLIRAssertion>> FactToDLIR(
+      Principal speaker, Fact fact) {
+    // TODO fill in real implementation. This is just a stub.
+    Predicate dummy_pred = Predicate("pred_name", {"arg1"}, kPositive);
+    DLIRAssertion dummy_assertion =
+        DLIRAssertion(DLIRCondAssertion(dummy_pred, {dummy_pred}));
+    // This vector needed to be declared before the call to make_pair because
+    // template argument inference could not figure it out otherwise.
+    std::vector<DLIRAssertion> dummy_assertions = {dummy_assertion};
+    return std::make_pair(dummy_pred, dummy_assertions);
+  }
+
+  // WIP need to figure out error message about the visitor being
+  // non-exhaustive even though it looks exhaustive AFAICT
+  std::vector<DLIRAssertion>
+  SingleSaysAssertionToDLIR(Principal speaker,
+      // BaseFact base_fact,
+      Assertion assertion) {
+
```

Comment:
@markww @bgogul Any idea why this doesn't work with AssertionVariantType? I get an error that this overloaded/visit is non-exhaustive even though `auto arg` should cover everything (also explicitly exhausting the 2 possibilities does not work either). This same overloaded/visit thing works just fine on BaseFact.GetValue() even if an implementation for just `auto arg` is given.

If this takes too long to look at, don't worry about it.

Reply:
Never mind, I figured this out (locally).